packages/query-core/src/queryClient.ts

```diff
@@ -105,32 +105,36 @@ export class QueryClient {
     return this.mutationCache.findAll({ ...filters, fetching: true }).length
   }
 
-  getQueryData<TData = unknown>(
+  getQueryData<TQueryFnData = unknown>(
     queryKey: QueryKey,
     filters?: QueryFilters,
-  ): TData | undefined {
-    return this.queryCache.find<TData>(queryKey, filters)?.state.data
+  ): TQueryFnData | undefined {
+    return this.queryCache.find<TQueryFnData>(queryKey, filters)?.state.data
   }
 
-  getQueriesData<TData = unknown>(queryKey: QueryKey): [QueryKey, TData][]
-  getQueriesData<TData = unknown>(filters: QueryFilters): [QueryKey, TData][]
-  getQueriesData<TData = unknown>(
+  getQueriesData<TQueryFnData = unknown>(
+    queryKey: QueryKey,
+  ): [QueryKey, TQueryFnData | undefined][]
+  getQueriesData<TQueryFnData = unknown>(
+    filters: QueryFilters,
+  ): [QueryKey, TQueryFnData | undefined][]
+  getQueriesData<TQueryFnData = unknown>(
     queryKeyOrFilters: QueryKey | QueryFilters,
-  ): [QueryKey, TData][] {
+  ): [QueryKey, TQueryFnData | undefined][] {
     return this.getQueryCache()
       .findAll(queryKeyOrFilters)
       .map(({ queryKey, state }) => {
-        const data = state.data as TData
+        const data = state.data as TQueryFnData
```

Comment:
the type assertion should be a union with undefined:

```suggestion
        const data = state.data as TQueryFnData | undefined
```